providers/rpc/rpc.go

```diff
@@ -0,0 +1,351 @@
+	SHA512 Algorithm = "sha512"
+	// SHA512Short is the short version of the SHA512 algorithm.
+	SHA512Short Algorithm = "512"
+)
+
+// Features implemented by the RPC provider.
+var Features = registrar.Features{
+	providers.FeaturePowerSet,
+	providers.FeaturePowerState,
+	providers.FeatureBootDeviceSet,
+}
+
+// Algorithm is the type for HMAC algorithms.
+type Algorithm string
+
+// Secrets hold per algorithm slice secrets.
+// These secrets will be used to create HMAC signatures.
+type Secrets map[Algorithm][]string
+
+// Signatures hold per algorithm slice of signatures.
+type Signatures map[Algorithm][]string
+
+// Config defines the configuration for sending rpc notifications.
+type Config struct {
+	// ConsumerURL is the URL where an rpc consumer/listener is running
+	// and to which we will send and receive all notifications.
+	ConsumerURL string
+	// Host is the BMC ip address or hostname or identifier.
+	Host string
+	// Logger is the logger to use for logging.
+	Logger logr.Logger
+	// LogNotificationsDisabled determines whether responses from rpc consumer/listeners will be logged or not.
+	LogNotificationsDisabled bool
+	// Opts are the options for the rpc provider.
+	Opts Opts
+
+	// listenerURL is the URL of the rpc consumer/listener.
+	listenerURL *url.URL
+}
+
+type Opts struct {
+	// Request is the options used to create the rpc HTTP request.
+	Request RequestOpts
+	// Signature is the options used for adding an HMAC signature to an HTTP request.
+	Signature SignatureOpts
+	// HMAC is the options used to create a HMAC signature.
+	HMAC HMACOpts
+	// Experimental options.
+	Experimental Experimental
+}
+
+type RequestOpts struct {
+	// Client is the http client used for all HTTP calls.
+	Client *http.Client
+	// HTTPContentType is the content type to use for the rpc request notification.
+	HTTPContentType string
+	// HTTPMethod is the HTTP method to use for the rpc request notification.
+	HTTPMethod string
+	// StaticHeaders are predefined headers that will be added to every request.
+	StaticHeaders http.Header
+	// TimestampFormat is the time format for the timestamp header.
+	TimestampFormat string
+	// TimestampHeader is the header name that should contain the timestamp. Example: X-BMCLIB-Timestamp
+	TimestampHeader string
+}
+
+type SignatureOpts struct {
+	// HeaderName is the header name that should contain the signature(s). Example: X-BMCLIB-Signature
+	HeaderName string
+	// AppendAlgoToHeaderDisabled decides whether to append the algorithm to the signature header or not.
+	// Example: X-BMCLIB-Signature becomes X-BMCLIB-Signature-256
+	// When set to true, a header will be added for each algorithm. Example: X-BMCLIB-Signature-256 and X-BMCLIB-Signature-512
+	AppendAlgoToHeaderDisabled bool
+	// IncludedPayloadHeaders are headers whose values will be included in the signature payload. Example: X-BMCLIB-My-Custom-Header
+	// All headers will be deduplicated.
+	IncludedPayloadHeaders []string
```

Comment:
How would this get used? Based on the example, it seems like a way to add arbitrary hedaers but I see we also have `StaticHeaders` elsewhere.

Reply:
I think this is a sort of filter that specifies which headers are included in the payload that gets signed?

The naming through me, but I think that's a me thing. Perhaps the example could be elaborated on 🤷🏻

Reply:
This is in the `SignatureOpts` and pertains only the the creation of the HMAC hash. These header names/keys must exist before hand. This option will not add headers to the request.

Reply:
in terms of the example, what do you think would improve the clarity of this field? maybe something like,

 "_given these headers in a request, the keys specified in `IncludedPayloadHeaders` will take the values of those keys and include them in the signature payload_"?

Reply:
updated